nixos/modules/services/web-apps/lemmy.nix

```diff
@@ -0,0 +1,142 @@
+        };
+        # TODO: this is a secret, should not be in nix-store
+        options.jwt_secret = mkOption {
+          type = types.str;
+          description = "json web token for authorization between server and client";
+        };
+        options.federation = {
+          enabled = mkEnableOption "whether to enable activitypub federation.";
+        };
+        options.captcha = {
+          enabled = mkOption {
+            type = types.bool;
+            default = true;
+            description = "Enable Captcha";
+          };
+          difficulty = mkOption {
+            type = types.str; # TODO: Types should be "easy", "medium", "hard". Define an enum.
+            default = "medium";
+            description = "The difficultly of the captcha to solve";
+          };
+        };
+      };
+      default = { };
+      description = "Lemmy configuration";
+    };
+
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.lemmy.settings = {
+      database.user = mkDefault "lemmy";
+      database.host = mkDefault "localhost";
+      database.port = mkDefault 5432;
+      database.database = mkDefault "lemmy";
+      database.pool_size = mkDefault 5;
+
+      bind = mkDefault "127.0.0.1";
+      tls_enabled = mkDefault true;
+      pictrs_url = mkDefault "http://localhost:8080";
+      actor_name_max_length = mkDefault 20;
+
+      rate_limit.message = mkDefault 180;
+      rate_limit.message_per_second = mkDefault 60;
+      rate_limit.post = mkDefault 6;
+      rate_limit.post_per_second = mkDefault 600;
+      rate_limit.register = mkDefault 3;
+      rate_limit.register_per_second = mkDefault 3600;
+      rate_limit.image = mkDefault 6;
+      rate_limit.image_per_second = mkDefault 3600;
+    };
+
+    # TODO: Check if this can't be put under /var/lib
+    environment.etc."lemmy.json".source =
+      settingsFormat.generate "config.hjson" cfg.settings;
+
+    services.postgresql.enable = true;
+
+    systemd.services.lemmy = {
+      environment = {
+        LEMMY_CONFIG_LOCATION = "/etc/lemmy.json";
+
+        # Verify how this is used, and don't put the password in the nix store
+        LEMMY_DATABASE_URL = with services.lemmy.settings.database;"postgres://${user}:${password}@${database}:${port}";
+      };
+
+      documentation = [
+        "https://join-lemmy.org/docs/en/administration/from_scratch.html"
+        "https://join-lemmy.org/docs"
+      ];
+
+      wantedBy = [
+        "multi-user.target"
+      ];
+
+      after = [
+        "lemmy-postgresql.service"
+      ];
+
+      requires = [
+        "lemmy-postgresql.service"
+      ];
+
+      serviceConfig = {
+        # TODO: verify that this is possible
+        DynamicUser = true;
+
+        # TODO: put the configuration here as well
+        StateDirectory = "lemmy";
+
+        ExecStart = "${pkgs.lemmy-server}/bin/lemmy_server";
+      };
+    };
+
+    # TODO: make this configurable
+    systemd.services.lemmy-postgresql = {
```

Comment:
Run this the `postgres` user to avoid `sudo`. Look at some other modules so see how others do it.

Reply:
The problem I ran into is that if you don't run it as the `lemmy` user, then `/var/lib/lemmy` does not exist.
Perhaps we need to add another step before to make sure the /var/lib/lemmy exists ?
Or potentially give up `DynamicUser`

Reply:
adding another one shot service to make sure the /var/lib/lemmy is there feels a little cumbersome. Would that justify a use of sudo here ?

Reply:
Oh ... good point.

Can we just execute a SQL statement to check if the database already exists instead?

Reply:
good idea, something like 
`psql -lqt | ${pkgs.coreutils}/bin/cut -d \| -f 1 | ${pkgs.gnugrep}/bin/grep -qw <db_name>`
should do.

Reply:
adressed

Reply:
Looks good!